staging/src/kubevirt.io/client-go/kubecli/websocket.go

```diff
@@ -55,57 +54,70 @@ func Dial(address string, tlsConfig *tls.Config) (*websocket.Conn, *http.Respons
 	return dialer.Dial(address, nil)
 }
 
-func Copy(dst *websocket.Conn, src *websocket.Conn) (written int64, err error) {
-	return copy(&binaryWriter{conn: dst}, &binaryReader{conn: src})
+func Copy(dst *websocket.Conn, src *websocket.Conn) (int64, error) {
+	return io.Copy(dst.UnderlyingConn(), src.UnderlyingConn())
 }
 
 func CopyFrom(dst io.Writer, src *websocket.Conn) (written int64, err error) {
-	return copy(dst, &binaryReader{conn: src})
+	return io.Copy(dst, &binaryReader{conn: src})
 }
 
 func CopyTo(dst *websocket.Conn, src io.Reader) (written int64, err error) {
-	return copy(&binaryWriter{conn: dst}, src)
-}
-
-func copy(dst io.Writer, src io.Reader) (written int64, err error) {
-	// our websocket package has an issue where it truncates messages
-	// when the message+header is greater than the buffer size we allocate.
-	// thus, we copy in chunks of WebsocketMessageBufferSize-wsFrameHeaderSize
-	buf := make([]byte, WebsocketMessageBufferSize-wsFrameHeaderSize)
-	return io.CopyBuffer(dst, src, buf)
+	return io.Copy(&binaryWriter{conn: dst}, src)
 }
 
 type binaryWriter struct {
-	conn *websocket.Conn
+	conn  *websocket.Conn
+	write int
 }
 
 func (s *binaryWriter) Write(p []byte) (int, error) {
+	s.write++
 	w, err := s.conn.NextWriter(websocket.BinaryMessage)
 	if err != nil {
 		return 0, convert(err)
 	}
 	defer w.Close()
-	return w.Write(p)
+	n, err := w.Write(p)
+	return n, err
 }
 
 type binaryReader struct {
-	conn *websocket.Conn
+	conn   *websocket.Conn
+	read   int
+	reader io.Reader
 }
 
 func (s *binaryReader) Read(p []byte) (int, error) {
+	var msgType int
+	var err error
 	for {
-		msgType, r, err := s.conn.NextReader()
+		s.read++
+		if s.reader == nil {
+			msgType, s.reader, err = s.conn.NextReader()
+		} else {
+			msgType = websocket.BinaryMessage
+		}
 		if err != nil {
+			s.reader = nil
 			return 0, convert(err)
 		}
 
 		switch msgType {
 		case websocket.BinaryMessage:
-			n, err := r.Read(p)
+			n, readErr := s.reader.Read(p)
+			err = readErr
+			if err != nil {
+				s.reader = nil
+				if err == io.EOF {
+					return 0, nil
```

Comment:
why not `return n, convert(err)` here?

Reply:
Because we don't know if additional messages will be there. We know at this stage only that the current binary message was read until the end. We have to at least try reading the next message. If there is no next message, we call `0, convert(err)`.

Reply:
why not use a 'continue' instead then here?

Reply:
would work. Let me do that.